packages/neon/neon/lib/src/models/push_notification.dart

```diff
@@ -1,28 +1,119 @@
+import 'dart:convert';
+
 import 'package:json_annotation/json_annotation.dart';
 import 'package:meta/meta.dart';
-import 'package:nextcloud/notifications.dart' show DecryptedSubject;
+import 'package:neon/src/utils/push_utils.dart';
+import 'package:nextcloud/notifications.dart' as notifications;
 
 part 'push_notification.g.dart';
 
-@JsonSerializable()
+/// The json key for [PushNotification.accountID].
+const String _accountIDKey = 'accountID';
+
+/// The json key for [PushNotification.priority].
+const String _priorityKey = 'priority';
+
+/// The json key for [PushNotification.type].
+const String _typeKey = 'type';
+
+/// The json key for [PushNotification.subject].
+const String _subjectKey = 'subject';
+
 @immutable
 @internal
-class PushNotification {
-  const PushNotification({
+abstract interface class PushNotification {
+  /// The account associated to this notification.
+  abstract final String accountID;
+
+  /// The priority of the notification.
+  abstract final String priority;
+
+  /// The type of the notification.
+  abstract final String type;
+
+  /// The subject of this notification.
+  abstract final notifications.DecryptedSubject subject;
+}
+
+/// Data for a push notification.
+@JsonSerializable()
+@internal
+class PushNotificationDecrypted implements PushNotification {
+  /// Creates a new push notification.
+  const PushNotificationDecrypted({
+    required this.accountID,
+    required this.priority,
+    required this.type,
+    required this.subject,
+  });
+
+  /// Creates a new PushNotification object from the given [json] data.
+  factory PushNotificationDecrypted.fromJson(final Map<String, dynamic> json) =>
+      _$PushNotificationDecryptedFromJson(json);
+
+  /// Parses this object into a json like map.
+  Map<String, dynamic> toJson() => _$PushNotificationDecryptedToJson(this);
+
+  @JsonKey(name: _accountIDKey)
+  @override
+  final String accountID;
+
+  @JsonKey(name: _priorityKey)
+  @override
+  final String priority;
+
+  @JsonKey(name: _typeKey)
+  @override
+  final String type;
+
+  @JsonKey(name: _subjectKey)
+  @override
+  final notifications.DecryptedSubject subject;
+}
+
+/// Data for a push notification.
+@JsonSerializable()
+@internal
+class PushNotificationEncrypted implements PushNotification {
+  /// Creates a new push notification.
+  PushNotificationEncrypted({
     required this.accountID,
     required this.priority,
     required this.type,
     required this.subject,
   });
 
-  factory PushNotification.fromJson(final Map<String, dynamic> json) => _$PushNotificationFromJson(json);
-  Map<String, dynamic> toJson() => _$PushNotificationToJson(this);
+  /// Creates a new PushNotification object from the given [json] data.
+  factory PushNotificationEncrypted.fromJson(
+    final Map<String, dynamic> json,
+  ) {
+    final privateKey = PushUtils.loadRSAKeypair();
```

Comment:
If you want to do it this way please pass in the private key instead of calling this logic in the model.